gno.land/pkg/sdk/vm/keeper.go

```diff
@@ -848,6 +870,113 @@ func (vm *VMKeeper) QueryDoc(ctx sdk.Context, pkgPath string) (*doc.JSONDocument
 	return d.WriteJSONDocumentation()
 }
 
+// QueryStorage returns storage and deposit for a realm.
+func (vm *VMKeeper) QueryStorage(ctx sdk.Context, pkgPath string) (string, error) {
+	store := vm.newGnoTransactionStore(ctx) // throwaway (never committed)
+	// Ensure pkgPath is realm.
+	if !gno.IsRealmPath(pkgPath) {
+		err := ErrInvalidPkgPath(fmt.Sprintf(
+			"package is not realm: %s", pkgPath))
+		return "", err
+	}
+
+	rlm := store.GetPackageRealm(pkgPath)
+	if rlm == nil {
+		err := ErrInvalidPkgPath(fmt.Sprintf(
+			"realm not found: %s", pkgPath))
+		return "", err
+	}
+	res := fmt.Sprintf("storage: %d, deposit: %d", rlm.Storage, rlm.Deposit)
+
+	return res, nil
+}
+
+// processDeposit processes storage deposit adjustments for package realms based on
+// storage size changes tracked within the gnoStore.
+//
+// For each realm, it:
+// - Charges the caller a deposit proportional to newly used storage (positive size difference).
+// - Returns the deposit to the caller for released storage (negative size difference).
+//
+// An error is returned if there's insufficient deposit, a transfer error occurs,
+// or an invariant violation happens during deposit or storage adjustments.
+
+func (vm *VMKeeper) processDeposit(ctx sdk.Context, caller crypto.Address, deposit std.Coins, gnostore gno.Store, params Params) error {
+	realmDiffs := gnostore.RealmDiffs()
+	depositAmt := deposit.AmountOf(ugnot.Denom)
+	if depositAmt == 0 {
+		depositAmt = std.MustParseCoin(params.DefaultDeposit).Amount
+	}
+	price := std.MustParseCoin(params.StoragePrice)
+
+	for rlmPath, diff := range realmDiffs {
+		if diff == 0 {
+			continue
+		}
+
+		rlm := gnostore.GetPackageRealm(rlmPath)
+
+		if diff > 0 {
+			// lock deposit for the additional storage used.
+			requiredDeposit := overflow.Mul64p(diff, price.Amount)
+			if depositAmt < requiredDeposit {
+				return fmt.Errorf("not enough deposit to cover the stroage usage, %d%s, %d bytes", requiredDeposit, ugnot.Denom, diff)
+			}
+			err := vm.lockStorageDeposit(ctx, caller, rlm, requiredDeposit, diff)
+			if err != nil {
+				return err
+			}
+		} else {
+			// release storage used and return deposit
+			released := -diff
+			if rlm.Storage < uint64(released) {
```

Comment:
note to self: why is this check needed? why is rlm.Storage needed?

Reply:
This sanity check prevents releasing more storage than was originally stored. It safeguards against future code changes that might break this invariant.